### PR TITLE
CLI: Introduce parameter presets, #158

### PR DESCRIPTION
**Pull request by kba**
The intention here is to simplify re-using parameter settings. Processors can bundle parameter setting presets, e.g. using a specific model or a specific combination of values. Users can store such presets at certain locations in the file system. A preset name can be used instead of a JSON string or filename for the `-p`/`--parameter` CLI option.

This is a first draft, bundled presets should probably described in `ocrd-tool.json` and lookup path simplified, etc.

### CODE REVIEW

**Comment by kba** [comment]
> bundled with a processor

means: stored as `package_data` and manageable with `pkg_resources`.

> But couldn't we handle --presets (listing) via normal --help?

Yes, that would be the human-readable version. `--presets` would only list the names of presets one-per-line for automation. I have a branch on the backburner that adds shell completion, that would be an application.

`sysconfig.get_path('data') will be equal to `$VIRTUAL_ENV` when using a venv/virtualenv (and we require it).

>  Perhaps we should first find a general convention for #160 and then just refer to that for preset files.

Fair point. I'll add a section on parameter filename resolution and we can agree on a search path first and then complete this PR.

**Comment by kba** [comment]
The file-parameter-resolution PR is now merged, I think this one is ready to ship as well (and then implement it in core for processors to make use of asap).

@bertsky Objections?

**Comment by bertsky** [comment]
> The file-parameter-resolution PR is now merged, I think this one is ready to ship as well (and then implement it in core for processors to make use of asap).
> 
> @bertsky Objections?

No, looks good!

Just that maybe we don't even need an extra `-C|--preset` now: If we generally require resolving relative path names for parameter files to adhere to the new ocrd-tool fpath conventions, then we can make the modules install their parameter files under these locations and users can just point to them via regular `-p|--parameter`. We would not even have to change any logic, only the documentation (including the term _preset_ itself). That would only leave `-L|--presets` as new. (But then again, why not also have a command to list all installed models under these paths?)

**Comment by kba** [comment]
We can use the same mechanism for presets and general file parameters, that is true (and should be reflected by referencing the resolution logic in ocrd_tool.md). *Preset* is limiting. "Bundled files" perhaps? "Included batteries" - something like that? Maybe just "assets"?

A generalized `-L/--list-bundled` (or whatever the wording) is still useful for automation, as is `-C/--dump-asset` (again: open on the wording). Having a mechanism short of actually running a processor to get to the files it bundles (or has available in the paths defined in the "file parameters" section in ocrd_tool), is still a sensible idea, again, for automation, generating user interfaces, things like dynamically loading a model into a memory-mapped FS location before processing etc.

**Comment by kba** [comment]
I've changed the text to be more generic. I chose `processor asset` as the working title.

**Comment by bertsky** [comment]
> We can use the same mechanism for presets and general file parameters, that is true

Sorry, I got confused. Of course we can – that's we have been planning. The `-C|--preset` was merely the dump action.

> (and should be reflected by referencing the resolution logic in ocrd_tool.md)

Perhaps the current text under the `Presets` section can be reformulated in a way a preset is nothing more than a relative filename for a bundled parameter file.

> _Preset_ is limiting. "Bundled files" perhaps? "Included batteries" - something like that? Maybe just "assets"?
> A generalized `-L/--list-bundled` (or whatever the wording) is still useful for automation, as is `-C/--dump-asset` (again: open on the wording). Having a mechanism short of actually running a processor to get to the files it bundles (or has available in the paths defined in the "file parameters" section in ocrd_tool), is still a sensible idea, again, for automation, generating user interfaces, things like dynamically loading a model into a memory-mapped FS location before processing etc.

Yes, absolutely. (I did not intend to argue against dump and list commands, see above.)

Okay, I get that it makes sense to subsume presets and models in both commands. One can still differentiate them by name / filename suffix I guess. 

For the wording, I would vote for _bundled files_, _resource files_, or _package resources_. _Assets_ is too much financial jargon to my taste (although I know it's become common via App programming). For example, `-L|--list-resources` and `-C|--dump-resource`.

File: cli.md [hunk 1]
```diff
@@ -168,14 +168,22 @@ The log messages must have the format `TIME LEVEL LOGGERNAME - MESSAGE\n`, where
...
+filename by following the [conventions laid out in the `ocrd_tool`
+spec](ocrd_tool#file-parameters). These files, either bundled by the processor
+developer or put in place by the user, are called *processor assets*. The
+*processor assets* of a processor can be listed with the `-L/--list-assets`
+option and individual *processor assets* can be retrieved with the
+`-C/--show-asset` option. Since *processor assets* use the same mechanism
+as file parameters, they can be used
+
+  * as the argument to the `-p/--parameter` option, and
+  * as the value of a file parameter
```
Comment (bertsky) [inline cli.md:183]:
```suggestion
  * as the argument to the `-p/--parameter` option (i.e. a **preset** file), and
  * as the value of a file-type parameter (e.g. a **model** file)
```

**Comment by kba** [comment]
> Assets is too much financial jargon to my taste (although I know it's become common via App programming)

The joy of a second language, I was not aware of either when creating `OCR-D/assets`...

I like `processor resources`.

**Pull request merged**